### PR TITLE
Math: Auditory: Fix the use of bitshift parameter

### PR DESCRIPTION
**Pull request by singalsu**
This fixes a large mismatch in SOF MFCC computation.

First patch is for auditory library to fix a log Mel equation, second patch adds tests those now detect the mistake.

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
@ShriramShastry @btian1 pls review

**Review by andrula-song — APPROVED** [review]

File: src/math/auditory/mel_filterbank_16.c [hunk 1]
```diff
@@ -74,9 +74,10 @@ void psy_apply_mel_filterbank_16(struct psy_mel_filterbank *fb, struct icomplex1
 		log += fb->scale_log2;
 
 		/* Change Q5.26 Q5.16 Subtract the applied lshift for power spectra
-		 * log2(x * 2^n) = log2(x) - n
+		 * log2(x * 2^n) = log2(x) - n. Note that the bitshift need to be subtracted
```
Comment (lyakh) [inline src/math/auditory/mel_filterbank_16.c:77]: shouldn't that be `2^-n` or `x / 2^n` or `log2(x) + n`?

File: src/math/auditory/mel_filterbank_32.c [hunk 2]
```diff
@@ -74,9 +74,10 @@ void psy_apply_mel_filterbank_32(struct psy_mel_filterbank *fb, struct icomplex3
 		log += fb->scale_log2;
 
 		/* Change Q5.26 Q5.16 Subtract the applied lshift for power spectra
-		 * log2(x * 2^n) = log2(x) - n
+		 * log2(x * 2^n) = log2(x) - n. Note that the bitshift need to be subtracted
```
Comment (lyakh) [inline src/math/auditory/mel_filterbank_32.c:77]: ditto

File: test/cmocka/src/math/auditory/mfcc_power_to_mel_log.m [hunk 3]
```diff
@@ -14,14 +15,14 @@
...
-function mel_log = mfcc_power_to_mel_log(state, param, s_half)
+function mel_log = mfcc_power_to_mel_log(state, param, s_half, shift)
 
 p_lin = s_half .* conj(s_half);
 mel_log = zeros(1, param.num_mel_bins);
 for i = 1:param.num_mel_bins
 	p = sum(p_lin .* state.triangles(:, i));
 	p = max(p, param.pmin);
-	mel_log(i) = log(p) * state.log_scale;
+	mel_log(i) = (log(p) - log(2^(2 * shift))) * state.log_scale;
```
Comment (lyakh) [inline test/cmocka/src/math/auditory/mfcc_power_to_mel_log.m:25]: not sure if it serves your purpose here, but you probably at least get a better precision if you do `2 * shift * log(2)`

File: src/math/auditory/mel_filterbank_16.c (hunk 1 above)
Comment (singalsu) [reply src/math/auditory/mel_filterbank_16.c:77]: Yep!

File: test/cmocka/src/math/auditory/mfcc_power_to_mel_log.m (hunk 3 above)
Comment (singalsu) [reply test/cmocka/src/math/auditory/mfcc_power_to_mel_log.m:25]: Probably more accurate, yes.  The reference is checked against Pytorch so I'll do this change even if it hides a bit why such equation.

**Review by lyakh — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**